miner/src/updater.rs

```diff
@@ -0,0 +1,78 @@
+use std::process::Command;
+use std::thread::{self, sleep};
+use std::time::Duration;
+use tracing::{info, error};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum UpdaterError {
+    #[error("Git pull failed: {0}")]
+    PullFailed(String),
+    
+    #[error("Build failed: {0}")]
+    BuildFailed(String),
+    
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+    
+    #[error(transparent)]
+    Utf8Error(#[from] std::string::FromUtf8Error)
+}
+
+pub struct Updater {
+    check_interval: Duration,
+    git_branch: String,
+}
+
+impl Updater {
+    pub fn new(check_interval: Duration, git_branch: String) -> Self {
+        Self { 
+            check_interval,
+            git_branch,
+        }
+    }
+
+    pub fn spawn(self) -> thread::JoinHandle<()> {
+        thread::spawn(move || {
+            loop {
+                if let Err(e) = self.check_and_update() {
+                    error!("Update check failed: {}", e);
+                }
+                sleep(self.check_interval);
+            }
+        })
+    }
+
+    fn check_and_update(&self) -> Result<(), UpdaterError> {
+        let pull = Command::new("git")
+            .args(["pull", "--ff-only", "origin", &self.git_branch])
+            .output()?;
+
+        if !pull.status.success() {
+            return Err(UpdaterError::PullFailed(
+                String::from_utf8_lossy(&pull.stderr).to_string()
+            ));
+        }
+
+        let stdout = String::from_utf8_lossy(&pull.stdout);
+        if stdout.contains("Already up to date") {
+            info!("No updates available");
+            return Ok(());
+        }
+
+        info!("Update found: {}", stdout.trim());
+
+        let build = Command::new("cargo")
+            .args(["run", "--release"])
+            .output()?;
+
+        if !build.status.success() {
+            return Err(UpdaterError::BuildFailed(
+                String::from_utf8_lossy(&build.stderr).to_string()
+            ));
+        }
+
+        info!("Update downloaded and built successfully. Restarting process to apply changes...");
+        Ok(())
+    }
+}
```

Comment:
Duplicating this code makes it harder to maintain. It should be moved to `base`.